auth/auth/auth.py

```diff
@@ -0,0 +1,233 @@
+        del session['session_id']
+
+    return web.HTTPFound(deploy_config.external_url('notebook2', ''))
+
+
+@routes.get('/api/v1alpha/login')
+async def rest_login(request):
+    callback_port = request.query['callback_port']
+
+    flow = get_flow(f'http://127.0.0.1:{callback_port}/oauth2callback')
+    authorization_url, state = flow.authorization_url(
+        access_type='offline',
+        include_granted_scopes='true')
+
+    return web.json_response({
+        'authorization_url': authorization_url,
+        'state': state
+    })
+
+
+@routes.get('/api/v1alpha/oauth2callback')
+async def rest_callback(request):
+    state = request.query['state']
+    code = request.query['code']
+    callback_port = request.query['callback_port']
+
+    try:
+        flow = get_flow(f'http://127.0.0.1:{callback_port}/oauth2callback', state=state)
+        flow.fetch_token(code=code)
+        token = google.oauth2.id_token.verify_oauth2_token(
+            flow.credentials.id_token, google.auth.transport.requests.Request())
+        id = token['sub']
+    except Exception:
+        log.exception('fetching and decoding token')
+        raise web.HTTPUnauthorized()
+
+    dbpool = request.app['dbpool']
+    async with dbpool.acquire() as conn:
+        async with conn.cursor() as cursor:
+            await cursor.execute('SELECT * from user_data where user_id = %s;', f'google-oauth2|{id}')
+            users = await cursor.fetchall()
+
+    if len(users) != 1:
+        raise web.HTTPUnauthorized()
+    user = users[0]
+
+    session_id = await create_session(dbpool, user['id'])
+
+    return web.json_response({
+        'token': session_id,
+        'username': user['username']
+    })
+
+
+@routes.post('/api/v1alpha/logout')
+@rest_authenticated_users_only
+async def rest_logout(request, userdata):
+    session_id = userdata['session_id']
+    dbpool = request.app['dbpool']
+    async with dbpool.acquire() as conn:
+        async with conn.cursor() as cursor:
+            await cursor.execute('DELETE FROM sessions WHERE session_id = %s;', session_id)
+
+    return web.Response(status=200)
+
+
+@routes.get('/api/v1alpha/userinfo')
+async def userinfo(request):
+    if 'Authorization' not in request.headers:
+        raise web.HTTPUnauthorized()
+
+    auth_header = request.headers['Authorization']
+    if not auth_header.startswith('Bearer '):
+        raise web.HTTPUnauthorized()
+    session_id = auth_header[7:]
+
+    # b64 encoding of 32-byte session ID is 44 bytes
+    if len(session_id) != 44:
+        raise web.HTTPUnauthorized()
+
+    dbpool = request.app['dbpool']
+    async with dbpool.acquire() as conn:
+        async with conn.cursor() as cursor:
+            await cursor.execute('''
+SELECT user_data.*, sessions.session_id FROM user_data
+INNER JOIN sessions ON user_data.id = sessions.user_id
+WHERE (sessions.session_id = %s) AND (ISNULL(sessions.max_age_secs) OR (NOW() < TIMESTAMPADD(SECOND, sessions.max_age_secs, sessions.created)));
```

Comment:
max_age_secs should never be null right? This logic also means that logout and session check use somewhat different semantics (logout deletes, but we'll have a bunch of dangling sessions because a user whose session is invalidated is effectively logged out, without actually calling the logout endpoint).

I think we could solve this by string an expired field on the session entry. (or I suppose deleting the session here).

Reply:
Yes, so I think the 2nd way to solve this dangling session issue is to changed _authenticated_users_only to always delete expired sessions. This could be done by forcing a redirect to /logout, but then we'd have dangling sessions with `@web_maybe_authenticated_users`

Reply:
> max_age_secs should never be null right?

Until we have key rotation in place, I was planning for the user user session secrets in K8s to have no expiration, hence explicitly supporting NULL max_age_secs.